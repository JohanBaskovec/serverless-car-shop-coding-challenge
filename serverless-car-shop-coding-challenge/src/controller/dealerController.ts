import { type CreateDealerDTO, type UpdateDealerDTO } from '../model/dto/dealerDTO.js';
import { CrudController } from './crudController.js';
import { type DealerDocument } from '../generated/mongoose.gen.js';

export class DealerController extends CrudController<DealerDocument, CreateDealerDTO, UpdateDealerDTO> {
}
