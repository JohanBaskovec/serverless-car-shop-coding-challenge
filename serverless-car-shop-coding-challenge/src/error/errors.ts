export class MissingRequestBody extends Error {
    constructor() {
        super('Missing body in HTTP request.');
        this.name = 'MissingRequestBody';
    }
}

export class MissingKeyInJSONBodyError extends Error {
    constructor(private readonly _missingKeys: string[]) {
        super(`Missing key in JSON: ${_missingKeys.join(',')}`);
        this.name = 'MissingKeyInJSONBodyError';
    }

    get missingKeys(): string[] {
        return this._missingKeys;
    }
}

export class MissingPathParametersError extends Error {
    constructor(private readonly _missingParameters: string[]) {
        super(`Missing path parameters: ${_missingParameters.join(',')}`);
        this.name = 'MissingPathParametersError';
    }

    get missingParameters(): string[] {
        return this._missingParameters;
    }
}

export class EntityDoesNotExistError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'EntityDoesNotExistError';
    }
}

export class ConstraintError extends Error {
    constructor(message: string) {
        super(message);
        this.name = 'ConstraintError';
    }
}
