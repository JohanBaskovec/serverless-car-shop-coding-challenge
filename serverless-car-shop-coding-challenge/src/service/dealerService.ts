import { Dealer } from '../model/dealer.js';
import { type CreateDealerDTO, type UpdateDealerDTO } from '../model/dto/dealerDTO.js';
import { ConstraintError, EntityDoesNotExistError, MissingKeyInJSONBodyError } from '../error/errors.js';
import mongoose from 'mongoose';
import { type VehicleService } from './vehicleService.js';
import { type DealerDocument, type VehicleDocument } from '../generated/mongoose.gen.js';
import { type CrudService } from './crudService.js';
import { type APIGatewayProxyEventQueryStringParameters } from 'aws-lambda/trigger/api-gateway-proxy.js';
import { type DeleteResponse } from '../model/response.js';

export class DealerService implements CrudService<DealerDocument, CreateDealerDTO, UpdateDealerDTO> {
    private _vehicleService: VehicleService;

    get vehicleService(): VehicleService {
        return this._vehicleService;
    }

    set vehicleService(value: VehicleService) {
        this._vehicleService = value;
    }

    async getAll(queryStringParameters: APIGatewayProxyEventQueryStringParameters | null = null): Promise<DealerDocument[]> {
        return await Dealer.find({}).exec();
    }

    async create(dto: CreateDealerDTO): Promise<DealerDocument> {
        const dealer = await new Dealer(dto).save();
        for (const vehicle of dealer.vehicles) {
            await this.vehicleService.setDealer(vehicle, dealer, false);
        }
        return dealer;
    }

    async update(dto: UpdateDealerDTO): Promise<DealerDocument> {
        if (dto._id == null) {
            throw new MissingKeyInJSONBodyError(['_id']);
        }
        if (dto.vehicles != null) {
            await this.setVehicles(
                new mongoose.Types.ObjectId(dto._id),
                dto.vehicles.map(v => new mongoose.Types.ObjectId(v)),
                true
            );
        }
        delete dto.vehicles;
        const dealer = await Dealer.findOneAndUpdate({ _id: dto._id }, dto, { new: true, runValidators: true }).exec();
        if (dealer == null) {
            throw new EntityDoesNotExistError(`Dealer with id ${dto._id} does not exist.`);
        }
        return dealer;
    }

    async remove(id: string): Promise<DeleteResponse> {
        const dealer: DealerDocument | null = await Dealer.findOne({ _id: id }).exec();
        if (dealer == null) {
            throw new EntityDoesNotExistError(`Dealer with id ${id} does not exist.`);
        }
        if ((dealer.vehicles as mongoose.Types.ObjectId[]).length !== 0) {
            throw new ConstraintError('Dealer has vehicles, delete them first before deleting the dealer.');
        }
        const response = await Dealer.deleteOne({ _id: id }).exec();
        if (response.deletedCount === 0) {
            throw new EntityDoesNotExistError(`Dealer with id ${id} does not exist.`);
        }
        return response;
    }

    async removeVehicle(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicle: mongoose.Types.ObjectId | VehicleDocument
    ): Promise<void> {
        const dealer = await this.getDealerDocument(dealerOrId);
        dealer.vehicles.pull(vehicle);
        await dealer.save();
    }

    async setVehicles(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehiclesToSet: Array<VehicleDocument | mongoose.Types.ObjectId>,
        updateOppositeSide = false
    ): Promise<void> {
        const dealer: DealerDocument = await this.getDealerDocument(dealerOrId);
        const vehiclesToAdd: Array<mongoose.Types.ObjectId | VehicleDocument> = [];
        // First we check if array contains new vehicles
        for (const vehicleToSet of vehiclesToSet) {
            let hasVehicle = false;
            for (const vehicle of dealer.vehicles) {
                if (vehicleToSet._id.equals(vehicle._id)) {
                    hasVehicle = true;
                    break;
                }
            }
            if (!hasVehicle) {
                vehiclesToAdd.push(vehicleToSet._id);
            }
        }
        // Then we check if any vehicle would be deleted from current list of vehicles
        for (const vehicle of dealer.vehicles) {
            let hasVehicle = false;
            for (const vehicleToSet of vehiclesToSet) {
                if (vehicleToSet._id.equals(vehicle._id)) {
                    hasVehicle = true;
                    break;
                }
            }
            if (!hasVehicle) {
                throw new ConstraintError(`Impossible to delete vehicle ${vehicle._id.toString()} from the dealer, as it wouldn't belong to a dealer anymore.`);
            }
        }
        dealer.vehicles.addToSet(...vehiclesToAdd);
        await dealer.save();
        if (updateOppositeSide) {
            for (const vehicle of vehiclesToAdd) {
                await this.vehicleService.setDealer(vehicle, dealer, false);
            }
        }
    }

    async addVehicles(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicles: Array<VehicleDocument | mongoose.Types.ObjectId>,
        updateOppositeSide = false
    ): Promise<void> {
        const dealer = await this.getDealerDocument(dealerOrId);
        const ids: mongoose.Types.ObjectId[] = [];
        for (const vehicle of vehicles) {
            if (vehicle instanceof mongoose.Types.ObjectId) {
                ids.push(vehicle);
            } else {
                ids.push(vehicle._id);
            }
        }
        dealer.vehicles.addToSet(ids);
        await dealer.save();
        if (updateOppositeSide) {
            for (const vehicle of vehicles) {
                await this.vehicleService.setDealer(vehicle, dealer, false);
            }
        }
    }

    async addVehicle(
        dealerOrId: DealerDocument | mongoose.Types.ObjectId,
        vehicle: VehicleDocument | mongoose.Types.ObjectId,
        updateOppositeSide = false
    ): Promise<void> {
        await this.addVehicles(dealerOrId, [vehicle], updateOppositeSide);
    }

    async getDealerDocument(dealerOrDealerId: DealerDocument | mongoose.Types.ObjectId): Promise<DealerDocument> {
        if (dealerOrDealerId instanceof mongoose.Types.ObjectId) {
            const newDealer = await Dealer.findOne({ _id: dealerOrDealerId }).exec();
            if (newDealer == null) {
                throw new EntityDoesNotExistError(`Dealer with id ${dealerOrDealerId.toString()} does not exist.`);
            }
            return newDealer;
        } else {
            return dealerOrDealerId;
        }
    }
}
