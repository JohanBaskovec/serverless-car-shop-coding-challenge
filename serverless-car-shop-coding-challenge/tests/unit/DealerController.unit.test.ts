import { DealerController } from '../../src/controller/dealerController.js';
import { type DealerService } from '../../src/service/dealerService.js';
import mongoose from 'mongoose';
import { EntityDoesNotExistError, MissingPathParametersError } from '../../src/error/errors.js';

describe('dealerController', () => {
    describe('getAll', () => {
        test('should return dealers', async() => {
            const service = {
                getAll: jest.fn().mockReturnValue([])
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.getAll();
            expect(service.getAll).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
    describe('create', () => {
        test('should create dealer when request is valid', async() => {
            const fakeNewDealer = {
                id: 'dealer-id',
                name: 'dealer-name'
            };
            const service = {
                create: jest.fn().mockImplementation(() => {
                    return fakeNewDealer;
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.create(JSON.stringify({}));
            expect(service.create).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
        test('should return an error when request is invalid', async() => {
            const service = {
                create: jest.fn().mockImplementation(() => {
                    throw new mongoose.Error.ValidationError();
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.create(JSON.stringify({}));
            expect(service.create).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
    describe('update', () => {
        test('should update dealer when request is valid', async() => {
            const fakeNewDealer = {
                id: 'dealer-id',
                name: 'dealer-name'
            };
            const service = {
                update: jest.fn().mockImplementation(() => {
                    return fakeNewDealer;
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.update(JSON.stringify({}));
            expect(service.update).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
        test('should return an error when request is invalid', async() => {
            const service = {
                update: jest.fn().mockImplementation(() => {
                    throw new mongoose.Error.ValidationError();
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.update(JSON.stringify({}));
            expect(service.update).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
    describe('remove', () => {
        test('should return an error when id is missing', async() => {
            const service = {
                remove: jest.fn().mockImplementation(() => {
                    throw new MissingPathParametersError(['id']);
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.remove({});
            expect(service.remove).toHaveBeenCalledTimes(0);
            expect(response).toMatchSnapshot();
        });
        test('should remove dealer when request is valid', async() => {
            const service = {
                remove: jest.fn().mockImplementation(() => {
                    return { name: 'response from Mango' };
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.remove({ id: 'an id' });
            expect(service.remove).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
        test("should return an error when dealer doesn't exist", async() => {
            const service = {
                remove: jest.fn().mockImplementation(() => {
                    throw new EntityDoesNotExistError("Dealer doesn't exist");
                })
            } as any as DealerService;
            const controller = new DealerController(service);
            const response = await controller.remove({ id: 'an id' });
            expect(service.remove).toHaveBeenCalledTimes(1);
            expect(response).toMatchSnapshot();
        });
    });
});
