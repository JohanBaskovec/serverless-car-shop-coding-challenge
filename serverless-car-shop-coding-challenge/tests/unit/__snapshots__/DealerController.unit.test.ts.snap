// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dealerController create should create dealer when request is valid 1`] = `
{
  "body": "{"data":{"id":"dealer-id","name":"dealer-name"}}",
  "statusCode": 200,
}
`;

exports[`dealerController create should return an error when request is invalid 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Validation failed"}}",
  "statusCode": 400,
}
`;

exports[`dealerController getAll should return dealers 1`] = `
{
  "body": "{"data":[]}",
  "statusCode": 200,
}
`;

exports[`dealerController remove should remove dealer when request is valid 1`] = `
{
  "body": "{"data":{"name":"response from Mango"}}",
  "statusCode": 200,
}
`;

exports[`dealerController remove should return an error when dealer doesn't exist 1`] = `
{
  "body": "{"error":{"type":"EntityDoesNotExistError","message":"Dealer doesn't exist"}}",
  "statusCode": 404,
}
`;

exports[`dealerController remove should return an error when id is missing 1`] = `
{
  "body": "{"error":{"type":"MissingPathParametersError","message":"Missing path parameters: id"}}",
  "statusCode": 400,
}
`;

exports[`dealerController update should return an error when request is invalid 1`] = `
{
  "body": "{"error":{"type":"ValidationError","message":"Validation failed"}}",
  "statusCode": 400,
}
`;

exports[`dealerController update should update dealer when request is valid 1`] = `
{
  "body": "{"data":{"id":"dealer-id","name":"dealer-name"}}",
  "statusCode": 200,
}
`;
